### PR TITLE
[FSDP] Refine utility APIs for FlatParam to get flat param sharding information

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75410
* __->__ #75248
* #75192
* #75537

1. This PR adds get_global_offset_from_param_offsets to map from a list of unflattened parameters offsets to the flattened parameter offsets.
2. This PR also unifies the naming and cleanup the classes used by FlatParam.

Differential Revision: [D35370411](https://our.internmc.facebook.com/intern/diff/D35370411/)

### CODE REVIEW

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 1]
```diff
@@ -74,13 +73,31 @@ def _pre_load_state_dict_hook(
...
     module_name: str
     module: nn.Module
     param_name: str
 
 
-class ShardMetadata(NamedTuple):
+class ParamMetadata(NamedTuple):
+    names: List[str]
+    shapes: List[torch.Size]
+    numels: List[int]
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:97]: do we need to keep track of `numels`? Can we get it from shape just by multiplying, or is it more convenient to have this attr?

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 2]
```diff
@@ -167,45 +190,98 @@ def __init__(self, params: Sequence[nn.Parameter], requires_grad: bool = True):
         # The indices (begin/end pair) of the parameters that are included in
         # this FlatParameter. The default value is all the parameters because
         # no sharding happen yet.
-        self._param_indice_in_shard = (0, len(self._param_infos) - 1)
+        self._param_indice_in_shard = Interval(0, len(self._param_infos) - 1)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flatten_params_wrapper.py:193]: seems ParamOffset is an alias for Interval, do we want to just use one or the other (i.e. change this to ParamOffset)?

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for adding this!

File: torch/distributed/fsdp/flatten_params_wrapper.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/flatten_params_wrapper.py:97]: We can infer numels from shapes. This is from the original code, and I'm not sure if meta is actually being used. Can do another refactor later.

**Pull request closed without merging**